drivers/xbee/xbee.c

```diff
@@ -0,0 +1,700 @@
+        DEBUG("xbee: Error sending data, payload length exceeds limit\n");
+        ng_pktbuf_release(pkt);
+        return 0;
+    }
+    /* get netif header check address length */
+    hdr = (ng_netif_hdr_t *)pkt->data;
+    if (!(hdr->dst_l2addr_len == 2 || hdr->dst_l2addr_len == 8)) {
+        ng_pktbuf_release(pkt);
+        return -ENOMSG;
+    }
+
+    /* acquire TX lock */
+    mutex_lock(&(dev->tx_lock));
+    /* put together the API frame */
+    dev->tx_buf[0] = API_START_DELIMITER;
+    dev->tx_buf[4] = 0;         /* set to zero to disable response frame */
+    /* set size, API id and address field depending on dst address length  */
+    if (hdr->dst_l2addr_len == 2) {
+        dev->tx_buf[1] = (uint8_t)((size + 5) >> 8);
+        dev->tx_buf[2] = (uint8_t)(size + 5);
+        dev->tx_buf[3] = API_ID_TX_SHORT_ADDR;
+        memcpy(dev->tx_buf + 5, ng_netif_hdr_get_dst_addr(hdr), 2);
+        pos = 7;
+    } else {
+        dev->tx_buf[1] = (uint8_t)((size + 11) >> 8);
+        dev->tx_buf[2] = (uint8_t)(size + 11);
+        dev->tx_buf[3] = API_ID_TX_LONG_ADDR;
+        memcpy(dev->tx_buf + 11, ng_netif_hdr_get_dst_addr(hdr), 8);
+        pos = 13;
+    }
+    /* set options */
+    dev->tx_buf[pos++] = dev->options;
+    /* copy payload */
+    payload = pkt->next;
+    while (payload) {
+        memcpy(&(dev->tx_buf[pos]), payload->data, payload->size);
+        pos += payload->size;
+        payload = payload->next;
+    }
+    /* set checksum */
+    dev->tx_buf[pos] = _cksum(dev->tx_buf, pos);
+    /* prepare transmission */
+    dev->tx_limit = (uint16_t)pos + 1;
+    dev->tx_count = 0;
+    /* start transmission */
+    uart_tx_begin(dev->uart);
+    /* release data */
+    ng_pktbuf_release(pkt);
+    /* return number of payload byte */
+    return (int)size;
+}
+
+int _add_cb(ng_netdev_t *dev, ng_netdev_event_cb_t cb)
+{
+    if (dev == NULL) {
+        return -ENODEV;
+    }
+    if (dev->event_cb != NULL) {
+        return -ENOBUFS;
+    }
+    dev->event_cb = cb;
+    return 0;
+}
+
+int _rem_cb(ng_netdev_t *dev, ng_netdev_event_cb_t cb)
+{
+    if (dev == NULL) {
+        return -ENODEV;
+    }
+    if (dev->event_cb != cb) {
+        return -ENOENT;
+    }
+    dev->event_cb = NULL;
+    return 0;
+}
+
+int _get(ng_netdev_t *netdev, ng_netconf_opt_t opt, void *value, size_t max_len)
+{
+    xbee_t *dev = (xbee_t *)netdev;
+    if (dev == NULL) {
+        return -ENODEV;
+    }
+
+    switch (opt) {
+        case NETCONF_OPT_ADDRESS:
+            return _get_addr_short(dev, (uint8_t *)value, max_len);
+        case NETCONF_OPT_ADDRESS_LONG:
+            return _get_addr_long(dev, (uint8_t *)value, max_len);
+        case NETCONF_OPT_CHANNEL:
+            return _get_channel(dev, (uint8_t *)value, max_len);
+        case NETCONF_OPT_NID:
+            return _get_panid(dev, (uint8_t *)value, max_len);
+        case NETCONF_OPT_PROTO:
+            return _get_proto(dev, (uint8_t *)value, max_len);
+        default:
```

Comment:
Okay… then I probably have to change my IPHC compression implementation a little :-D